contrib/camel/pkg/reconciler/camelsource.go

```diff
@@ -0,0 +1,326 @@
+			source.Status.MarkDeploying("IntegrationUpdated", "Updated integration %s", integration.Name)
+		} else {
+			source.Status.MarkDeploying("IntegrationNeedsUpdate", "Attempting to update integration %s", integration.Name)
+			r.recorder.Eventf(source, corev1.EventTypeWarning, "IntegrationNeedsUpdate", "Failed to update integration %q", integration.Name)
+		}
+		// Return after this update or error and reconcile again
+		return nil, err
+	}
+	return integration, nil
+}
+
+func (r *reconciler) getIntegration(ctx context.Context, source *v1alpha1.CamelSource) (*camelv1alpha1.Integration, error) {
+	logger := logging.FromContext(ctx)
+
+	list := &camelv1alpha1.IntegrationList{}
+	err := r.client.List(
+		ctx,
+		&client.ListOptions{
+			Namespace:     source.Namespace,
+			LabelSelector: labels.Everything(),
+			// TODO this is here because the fake client needs it.
+			// Remove this when it's no longer needed.
+			Raw: &metav1.ListOptions{
+				TypeMeta: metav1.TypeMeta{
+					APIVersion: camelv1alpha1.SchemeGroupVersion.String(),
+					Kind:       "Integration",
+				},
+			},
+		},
+		list)
+	if err != nil {
+		logger.Errorw("Unable to list integrations", zap.Error(err))
+		return nil, err
+	}
+	for _, c := range list.Items {
+		if metav1.IsControlledBy(&c, source) {
+			return &c, nil
+		}
+	}
+	return nil, k8serrors.NewNotFound(schema.GroupResource{}, "")
+}
+
+func (r *reconciler) createIntegration(ctx context.Context, source *v1alpha1.CamelSource, args *resources.CamelArguments) (*camelv1alpha1.Integration, error) {
+	integration, err := resources.MakeIntegration(args)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := controllerutil.SetControllerReference(source, integration, r.scheme); err != nil {
+		return nil, err
+	}
+
+	if err := r.client.Create(ctx, integration); err != nil {
+		return nil, err
+	}
+	return integration, nil
+}
+
+func (r *reconciler) reconcilePlatform(ctx context.Context, namespace string) error {
+	_, err := r.getPlatform(ctx, namespace)
+	if err != nil && k8serrors.IsNotFound(err) {
+		// Create a platform with default configuration if not present
+		platform := resources.MakePlatform(namespace)
+		return r.client.Create(ctx, platform)
+	}
+	return err
+}
+
+func (r *reconciler) getPlatform(ctx context.Context, namespace string) (*camelv1alpha1.IntegrationPlatform, error) {
+	platform := camelv1alpha1.IntegrationPlatform{}
+	key := client.ObjectKey{
+		Namespace: namespace,
+		Name:      "camel-k",
+	}
+	if err := r.client.Get(ctx, key, &platform); err != nil {
+		return nil, err
+	}
+	return &platform, nil
+}
+
+func (r *reconciler) reconcileIntegrationContext(ctx context.Context, namespace string, image string) (*camelv1alpha1.IntegrationContext, error) {
+	ct, err := r.getIntegrationContext(ctx, namespace, image)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Create a context with default configuration if not present
+			ct = resources.MakeContext(namespace, image)
```

Comment:
Normally we would want to set the owner on the new context object that was a result of the source resource. But in the case of camel, I am not sure. Please advise.

Reply:
Only the Integration resource needs to be owned in the case of Camel K

Reply:
What cleans up old IntegrationContexts once there are no more sources which reference the `image`?

Reply:
There's a garbage collector mechanism in Camel K that removes contexts that are no longer referenced.